### PR TITLE
Add cgroup v2 support to Interop.cgroups

### PR DESCRIPTION
**Pull request by omajid**
Fix up code to adjust cgroup v1 assumptions and check cgroup v2 paths, locations and values.

Continue using the older cgroup v1 terminology for APIs.

### CODE REVIEW

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 1]
```diff
@@ -79,6 +85,16 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
                         result = checked(ulongValue * multiplier);
                         return true;
                     }
+
+                    var encoding = new UTF8Encoding();
+                    string value = encoding.GetString(bytes);
+                    if ("max".Equals(value, StringComparison.InvariantCultureIgnoreCase))
```
Comment (omajid) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: cgroupv2 uses explicit values or `max` to indicate there's no memory limit active. For example:

```
$ cat /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/memory.max 
max
```

This code here treats that as not having a memory limit defined, so the caller (`Process.GetWorkingSetLimits`) can fall back to reading `/proc/self/stat`.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 2]
```diff
@@ -125,7 +166,7 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
                         while ((line = reader.ReadLine()) != null)
                         {
                             // Look for an entry that has cgroup as the "filesystem type"
-                            // and that has options containing the specified subsystem.
+                            // and, for cgroup1, that has options containing the specified subsystem
```
Comment (omajid) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: For cgroup v2, no subsystem is included in the mount info; there's only a single global mount point. For example:

```
$ cat /proc/self/mountinfo | grep cgroup2
27 20 0:24 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel,nsdelegate
```

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 3]
```diff
@@ -113,7 +154,7 @@ private static string FindCGroupPath(string subsystem)
         /// <param name="root">The path of the directory in the filesystem which forms the root of this mount; null if not found.</param>
         /// <param name="path">The path of the mount point relative to the process's root directory; null if not found.</param>
         /// <returns>true if the mount was found; otherwise, null.</returns>
-        private static bool TryFindHierarchyMount(string subsystem, out string root, out string path)
+        private static bool TryFindHierarchyMount(string subsystem, out CGroupVersion version, out string root, out string path)
```
Comment (omajid) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: We need to know the version of cgroup because that affects the information we are looking for when parsing `/proc/self/cgroup` later on.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 4]
```diff
@@ -178,37 +236,56 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
...
+                                return true;
                             }
+                        }
+                        else if (version == CGroupVersion.CGroup2)
+                        {
+                            string line;
+                            while ((line = reader.ReadLine()) != null)
+                            {
+                                // Find the first entry that matches the cgroup v2 hierarchy:
+                                //     0::$PATH
```
Comment (omajid) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: For example:

```
$ cat /proc/self/cgroup 
0::/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
```

File: src/System.Runtime.InteropServices.RuntimeInformation/tests/DescriptionNameTests.cs [hunk 5]
```diff
@@ -40,7 +40,7 @@ public void DumpRuntimeInformationToConsole()
 
             Console.WriteLine($"### CURRENT DIRECTORY: {Environment.CurrentDirectory}");
 
-            string cgroupsLocation = Interop.cgroups.s_cgroupMemoryPath;
+            string cgroupsLocation = Interop.cgroups.s_cgroupMemoryLimitPath;
```
Comment (omajid) [inline src/System.Runtime.InteropServices.RuntimeInformation/tests/DescriptionNameTests.cs]: Test output example on my system:

<details>
        <summary>Click to expand</summary>

```
  ### OS: Distro=Distro=fedora VersionId=30.0 Description=Linux 5.1.18-300.fc30.x86_64 #1 SMP Mon Jul 15 15:42:34 UTC 2019 Version=Unix 5.1.18.300 Arch=X64
  ### LIBC: Release=stable Version=2.29
  ### FRAMEWORK: Version=5.0.0 Description=.NET Core 5.0.0
  ### BINARIES: /home/omajid/devel/dotnet/corefx/artifacts/bin/testhost/netcoreapp-Linux-Debug-x64/shared/Microsoft.NETCore.App/5.0.0 (drive format ext4)
  ### TEMP PATH: /tmp/ (drive format tmpfs)
  ### CURRENT DIRECTORY: /home/omajid/devel/dotnet/corefx/artifacts/bin/System.Runtime.InteropServices.RuntimeInformation.Tests/netcoreapp-Unix-Debug
  ### CGROUPS MEMORY: /sys/fs/cgroup/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service/memory.max
  ### ENVIRONMENT VARIABLES
<removed>
  ### PROCESS INFORMATION:
  ###   Architecture: X64
  ###   BasePriority: 0
  ###   HandleCount: 177
  ###   Id: 23194
  ###   MachineName: .
  ###   MainModule: System.Diagnostics.ProcessModule (dotnet)
  ###   MainWindowHandle: 0
  ###   MainWindowTitle:
  ###   MaxWorkingSet: 9223372036854775807
  ###   MinWorkingSet: 0
  ###   NonpagedSystemMemorySize: 0
  ###   NonpagedSystemMemorySize64: 0
  ###   PagedMemorySize: 0
  ###   PagedMemorySize64: 0
  ###   PagedSystemMemorySize: 0
  ###   PagedSystemMemorySize64: 0
  ###   PeakPagedMemorySize: 0
  ###   PeakPagedMemorySize64: 0
  ###   PeakVirtualMemorySize: 0
  ###   PeakVirtualMemorySize64: 0
  ###   PeakWorkingSet: 0
  ###   PeakWorkingSet64: 0
  ###   PriorityBoostEnabled: False
  ###   PriorityClass: Normal
  ###   PrivateMemorySize: 0
  ###   PrivateMemorySize64: 0
  ###   PrivilegedProcessorTime: 00:00:00.0300000
  ###   ProcessName: dotnet
  ###   ProcessorAffinity: 255
  ###   Responding: True
  ###   SessionId: 10889
  ###   StartTime: 7/22/2019 10:07:42 PM
  ###   TotalProcessorTime: 00:00:00.9200000
  ###   UserProcessorTime: 00:00:00.8900000
  ###   VirtualMemorySize: -471937024
  ###   VirtualMemorySize64: 3823030272
  ###   WorkingSet: 53080064
  ###   WorkingSet64: 53080064

  ### CONTENTS OF "/proc/self/mountinfo":
  20 57 0:20 / /sys rw,nosuid,nodev,noexec,relatime shared:2 - sysfs sysfs rw,seclabel
  21 57 0:4 / /proc rw,nosuid,nodev,noexec,relatime shared:13 - proc proc rw
  22 57 0:6 / /dev rw,nosuid shared:9 - devtmpfs devtmpfs rw,seclabel,size=7905916k,nr_inodes=1976479,mode=755
  23 20 0:7 / /sys/kernel/security rw,nosuid,nodev,noexec,relatime shared:3 - securityfs securityfs rw
  24 22 0:21 / /dev/shm rw,nosuid,nodev shared:10 - tmpfs tmpfs rw,seclabel
  25 22 0:22 / /dev/pts rw,nosuid,noexec,relatime shared:11 - devpts devpts rw,seclabel,gid=5,mode=620,ptmxmode=000
  26 57 0:23 / /run rw,nosuid,nodev shared:12 - tmpfs tmpfs rw,seclabel,mode=755
  27 20 0:24 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:4 - cgroup2 cgroup2 rw,seclabel,nsdelegate
  28 20 0:25 / /sys/fs/pstore rw,nosuid,nodev,noexec,relatime shared:5 - pstore pstore rw,seclabel
  29 20 0:26 / /sys/fs/bpf rw,nosuid,nodev,noexec,relatime shared:6 - bpf bpf rw,mode=700
  52 20 0:27 / /sys/kernel/config rw,relatime shared:7 - configfs configfs rw
  57 0 253:0 / / rw,relatime shared:1 - ext4 /dev/mapper/luks-deadbeef-dead-beef-dead-deadbeef rw,seclabel
  30 20 0:19 / /sys/fs/selinux rw,relatime shared:8 - selinuxfs selinuxfs rw
  31 20 0:8 / /sys/kernel/debug rw,relatime shared:14 - debugfs debugfs rw,seclabel
  32 21 0:30 / /proc/sys/fs/binfmt_misc rw,relatime shared:15 - autofs systemd-1 rw,fd=37,pgrp=1,timeout=0,minproto=5,maxproto=5,direct,pipe_ino=16196
  33 57 0:31 / /tmp rw,nosuid,nodev shared:16 - tmpfs tmpfs rw,seclabel
  34 22 0:18 / /dev/mqueue rw,relatime shared:17 - mqueue mqueue rw,seclabel
  35 22 0:32 / /dev/hugepages rw,relatime shared:18 - hugetlbfs hugetlbfs rw,seclabel,pagesize=2M
  36 20 0:33 / /sys/fs/fuse/connections rw,relatime shared:19 - fusectl fusectl rw
  80 57 8:1 / /boot rw,relatime shared:39 - ext4 /dev/sda1 rw,seclabel
  83 57 253:1 / /home rw,relatime shared:41 - ext4 /dev/mapper/luks-deadbeef-dead-beef-dead-deadbeef rw,seclabel
  174 57 0:34 / /var/lib/nfs/rpc_pipefs rw,relatime shared:87 - rpc_pipefs sunrpc rw
  559 26 0:37 / /run/user/1000 rw,nosuid,nodev,relatime shared:308 - tmpfs tmpfs rw,seclabel,size=1584160k,mode=700,uid=1000,gid=1000
  349 559 0:35 / /run/user/1000/gvfs rw,nosuid,nodev,relatime shared:147 - fuse.gvfsd-fuse gvfsd-fuse rw,user_id=1000,group_id=1000
  ### CONTENTS OF "/proc/self/cgroup":
  0::/user.slice/user-1000.slice/user@1000.service/gnome-terminal-server.service
  ### CONTENTS OF "/proc/self/limits":
  Limit                     Soft Limit           Hard Limit           Units
  Max cpu time              unlimited            unlimited            seconds
  Max file size             unlimited            unlimited            bytes
  Max data size             unlimited            unlimited            bytes
  Max stack size            8388608              unlimited            bytes
  Max core file size        unlimited            unlimited            bytes
  Max resident set          unlimited            unlimited            bytes
  Max processes             61764                61764                processes
  Max open files            524288               524288               files
  Max locked memory         65536                65536                bytes
  Max address space         unlimited            unlimited            bytes
  Max file locks            unlimited            unlimited            locks
  Max pending signals       61764                61764                signals
  Max msgqueue size         819200               819200               bytes
  Max nice priority         0                    0
  Max realtime priority     0                    0
  Max realtime timeout      unlimited            unlimited            us
    Finished:    System.Runtime.InteropServices.RuntimeInformation.Tests
  === TEST EXECUTION SUMMARY ===
     System.Runtime.InteropServices.RuntimeInformation.Tests  Total: 8, Errors: 0, Failed: 0, Skipped: 0, Time: 0.199s
  ~/devel/dotnet/corefx/src/System.Runtime.InteropServices.RuntimeInformation/tests
  ----- end Mon 22 Jul 2019 10:07:44 PM EDT ----- exit code 0 ----------------------------------------------------------
  exit code 0 means Exited Successfully
```
</details>

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 6]
```diff
@@ -90,12 +106,37 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
...
         /// <returns>The cgroup path if found; otherwise, null.</returns>
-        private static string FindCGroupPath(string subsystem)
+        private static string FindCGroupPath(string subsystem, out CGroupVersion version)
         {
-            if (TryFindHierarchyMount(subsystem, out string hierarchyRoot, out string hierarchyMount) &&
+            version = CGroupVersion.None;
+            if (TryFindHierarchyMount(subsystem, out version, out string hierarchyRoot, out string hierarchyMount) &&
                 TryFindCGroupPathForSubsystem(subsystem, out string cgroupPathRelativeToMount))
             {
                 // For a host cgroup, we need to append the relative path.
```
Comment (omajid) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: I don't have a container runtime that currently supports cgroup v2 so I can't verify this host vs container difference in cgroup v2.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 7]
```diff
@@ -79,6 +85,16 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
                         result = checked(ulongValue * multiplier);
                         return true;
                     }
+
+                    var encoding = new UTF8Encoding();
+                    string value = encoding.GetString(bytes);
```
Comment (bartonjs) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: The only difference between `new UTF8Encoding()` and `Encoding.UTF8` is whether GetPreamble() returns empty or non-empty.  Since this isn't calling GetPreamble, or using a TextWriter, presumably it can use the cached object instead of allocating a new one?

Comment (bartonjs) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Also, maybe we want to just save that to a static initializer span and compare the bytes (if it's guaranteed to be UTF-8/ASCII).

Unless this is all one-time startup code, in which case maybe it's not worth it.

File: src/Common/tests/Tests/Interop/cgroupsTests.cs [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+        [InlineData(true, "1k",  1024)]
+        [InlineData(true, "1K",  1024)]
+        public static void ValidateTryReadMemoryValue(bool expectedResult, string valueText, ulong expectedValue)
+        {
+            string path = Path.GetTempFileName();
+            try
+            {
+                File.WriteAllText(path, valueText);
+
+                var result = Interop.cgroups.TryReadMemoryValueFromFile(path, out ulong val);
```
Comment (bartonjs) [inline src/Common/tests/Tests/Interop/cgroupsTests.cs]: Style: Improper var.  (I'm guessing `bool`?) (Repeats 3x)

File: src/Common/tests/Tests/Interop/cgroupsTests.cs [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+                Assert.Equal(expectedResult, result);
+                if (result)
+                {
+                    Assert.Equal(expectedValue, val);
+                }
+            }
+            finally {
```
Comment (bartonjs) [inline src/Common/tests/Tests/Interop/cgroupsTests.cs]: Style: Open brace should be on the next line (repeats 3x).

File: src/Common/tests/Tests/Interop/cgroupsTests.cs (hunk 8 above)
Comment (omajid) [reply src/Common/tests/Tests/Interop/cgroupsTests.cs]: Fixed now.

File: src/Common/tests/Tests/Interop/cgroupsTests.cs (hunk 9 above)
Comment (omajid) [reply src/Common/tests/Tests/Interop/cgroupsTests.cs]: Thanks. Fixed now.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 7 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: The call chain is:

- `Interop.cgroups.TryGetMemoryLimit`
- `Process.GetWorkingSetLimits`
- `Process.MaxWorkingSet.get`

It seems to me that `MaxWorkingSet.get` can only call this method one-time: https://github.com/dotnet/corefx/blob/c265373119bc0449d347862862b43de35a95b186/src/System.Diagnostics.Process/src/System/Diagnostics/Process.cs#L262-L273

**Comment by omajid** [comment]
cc @tmds

**Comment by danmosemsft** [comment]
cc @janvorli

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 10]
```diff
@@ -148,17 +193,29 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
                                 continue;
                             }
 
-                            if (postSeparatorlineParts[0] != "cgroup" ||
-                                Array.IndexOf(postSeparatorlineParts[2].Split(','), subsystem) < 0)
+                            if ((postSeparatorlineParts[0] == "cgroup" &&
+                                 Array.IndexOf(postSeparatorlineParts[2].Split(','), subsystem) < 0) &&
+                                postSeparatorlineParts[0] != "cgroup2")
```
Comment (omajid) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: This condition is wrong. I am working on a fix.

Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Fixed now. Also added a unit test to exercise it.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 1 above)
Comment (stephentoub) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Can this be OrdinalIgnoreCase rather than InvariantCultureIgnoreCase?

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 11]
```diff
@@ -79,6 +85,15 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
                         result = checked(ulongValue * multiplier);
                         return true;
                     }
+
+                    string value = Encoding.UTF8.GetString(bytes);
+                    if ("max".Equals(value, StringComparison.InvariantCultureIgnoreCase))
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: I don't understand what this check is doing.  If Equals returns true, we set result to 0 and return false.  If Equals returns false, we'll fall through to the exist code path that sets result to 0 and returns false.  So... why do we need this extra check at all?

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 12]
```diff
@@ -79,6 +85,15 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
                         result = checked(ulongValue * multiplier);
                         return true;
                     }
+
+                    string value = Encoding.UTF8.GetString(bytes);
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: We shouldn't need to get the string here; we can just use a span-based comparison on the bytes themselves.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 13]
```diff
@@ -90,17 +105,42 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
...
+            if (cgroupMemoryPath == null)
+            {
+                return null;
+            }
+
+            if (version == CGroupVersion.CGroup1)
+            {
+                return cgroupMemoryPath + "/memory.limit_in_bytes";
+            }
+            else if (version == CGroupVersion.CGroup2)
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Nit: this `else` isn't necessary, and it'd better follow the rest of the method's pattern if this were instead just an `if` with a blank line above it

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 14]
```diff
@@ -90,17 +105,42 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
...
+            {
+                // 'memory.high' is a soft limit; the process may get throttled
+                // 'memory.max' is where OOM killer kicks in
+                return cgroupMemoryPath + "/memory.max";
+            }
+
+            return null;
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: You could avoid having two `return null;`s if you inverted the `if (cgroupMemoryPath == null)` above.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 15]
```diff
@@ -90,17 +105,42 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
...
+        }
+
         /// <summary>Find the cgroup path for the specified subsystem.</summary>
         /// <param name="subsystem">The subsystem, e.g. "memory".</param>
         /// <returns>The cgroup path if found; otherwise, null.</returns>
-        private static string FindCGroupPath(string subsystem)
+        private static string FindCGroupPath(string subsystem, out CGroupVersion version)
         {
-            if (TryFindHierarchyMount(subsystem, out string hierarchyRoot, out string hierarchyMount) &&
+            version = CGroupVersion.None;
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: This line should not be necessary.  The first thing the method does after this is call TryFindHierarchyMount, which takes `out version`.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 16]
```diff
@@ -90,17 +105,42 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
...
+            if (TryFindHierarchyMount(subsystem, out version, out string hierarchyRoot, out string hierarchyMount) &&
                 TryFindCGroupPathForSubsystem(subsystem, out string cgroupPathRelativeToMount))
             {
                 // For a host cgroup, we need to append the relative path.
                 // In a docker container, the root and relative path are the same and we don't need to append.
-                return (hierarchyRoot != cgroupPathRelativeToMount) ?
+                return !string.Equals(hierarchyRoot, cgroupPathRelativeToMount, StringComparison.Ordinal) ?
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Why was this change necessary? `!=` will do an ordinal comparison.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 17]
```diff
@@ -148,17 +193,32 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
                                 continue;
                             }
 
-                            if (postSeparatorlineParts[0] != "cgroup" ||
-                                Array.IndexOf(postSeparatorlineParts[2].Split(','), subsystem) < 0)
+                            if (!string.Equals(postSeparatorlineParts[0], "cgroup2", StringComparison.Ordinal) &&
+                                (!string.Equals(postSeparatorlineParts[0], "cgroup", StringComparison.Ordinal) ||
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: As above, I'm not clear on why the explicit `Equals` call with `Ordinal` is better than just using `!=`.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 18]
```diff
@@ -148,17 +193,32 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
...
+                                  Array.IndexOf(postSeparatorlineParts[2].Split(','), subsystem) < 0))
                             {
                                 // Not the relevant entry.
                                 continue;
                             }
 
-                            // Found the relevant entry.  Extract the mount root and path.
+                            // Found the relevant entry.  Extract the cgroup version, mount root and path.
+                            version = CGroupVersion.None;
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: This line can be moved into the default branch below.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 19]
```diff
@@ -148,17 +193,32 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
...
+                                case "cgroup":
+                                    version = CGroupVersion.CGroup1;
+                                    break;
+                                case "cgroup2":
+                                    version = CGroupVersion.CGroup2;
+                                    break;
+                                default:
+                                    Debug.Fail($"invalid value for CGroupVersion \"{postSeparatorlineParts[0]}\"");
+                                    break;
+                            }
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Nit: please add a blank line below this brace

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 20]
```diff
@@ -180,27 +241,42 @@ private static bool TryFindHierarchyMount(string subsystem, out string root, out
...
                             if (lineParts.Length != 3)
                             {
                                 // Malformed line.
                                 continue;
                             }
 
+                            // cgroup v2: Find the first entry that matches the cgroup v2 hierarchy:
+                            //     0::$PATH
+
+                            if (lineParts[0] == "0" && string.Equals(string.Empty, lineParts[1], StringComparison.Ordinal))
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Same comparison question

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Thanks, @bartonjs pointed this out too. This should be fixed in the new commit.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 11 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Strictly speaking, we don't. I figured I would handle the entire set of valid input values here explicitly. `max` is the only non-numeric value possible and I wanted to handle it in the body along with all other numeric values.

Should I add a check for other values and warn if anything other than a numeric value and `max` is seen?

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 1 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Yes. We can probably even do plain `Ordinal`, `max` is the only stringthat the specification asks for.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 12 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: @stephentoub do you mean something like converting the string literal `max` to a byte array (using a decoding) and then doing `bytes.AsSpan() == max.AsSpan()`?

Comment (stephentoub) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Yes, e.g.
```C#
private static ReadOnlySpan<byte> MaxBytes => new byte[] { (byte)'m', (byte)'a', (byte)'x' };
...
if (MaxBytes.Equals(value, StringComparison.InvariantCultureIgnoreCase))
{
   ...
}
```
(While that MaxBytes looks like it's going to allocate every time, if you look at the IL you'll see the C# compiler doesn't actually do that.)

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 21]
```diff
@@ -79,6 +87,14 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
                         result = checked(ulongValue * multiplier);
                         return true;
                     }
+
+                    if (MaxBytes == bytes.AsSpan())
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: This unfortunately is not going to do what you want it to do, as == on span behaves differently than == on string.  This one could use Equals with StringComparison, but it'd be better just as:
```C#
if (MaxBytes.SequenceEquals(bytes))
```

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 22]
```diff
@@ -79,6 +87,14 @@ private static bool TryReadMemoryValueFromFile(string path, out ulong result)
...
+                    {
+                        // This is not (UIn64.MaxValue, true) because we want to let the caller
+                        // fallback to reading the real limits via other means
+                        result = 0;
+                        return false;
+                    }
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Per my previous comments, I would just delete this whole if block and let it fall through to the `result = 0; return false;` below.  If you want to add a comment about "max", that's fine, but this code is just overhead for no actual benefit.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs [hunk 23]
```diff
@@ -38,11 +44,13 @@ public static bool TryGetMemoryLimit(out ulong limit)
             return false;
         }
 
+        private static ReadOnlySpan<byte> MaxBytes => new byte[] { (byte)'m', (byte)'a', (byte)'x' };
```
Comment (stephentoub) [inline src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: If the below if block I commented on ends up being the only use of this, then this can be deleted, too.

**Comment by omajid** [comment]
@stephentoub Thanks for the review and your great suggestions! Does the updated patch look okay now?

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 23 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Removed.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 22 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Okay. I added a comment and removed the entire check.

File: src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs (hunk 16 above)
Comment (omajid) [reply src/Common/src/Interop/Linux/cgroups/Interop.cgroups.cs]: Sorry. I have a java background, sometimes that seeps through. These should all be fixed now.

**Comment by stephentoub** [comment]
LGTM. Thanks.  I pushed a commit to clean up the tests a bit, in particular to use our FileCleanupTestBase helper rather than generating random file names and tracking deleting them.

**Comment by omajid** [comment]
> I pushed a commit to clean up the tests a bit, in particular to use our FileCleanupTestBase helper rather than generating random file names and tracking deleting them.

Thanks, that's very kind of you!

**Comment by tmds** [comment]
The related PR in coreclr is needed to make the runtime/GC aware of cgroup v2: https://github.com/dotnet/coreclr/pull/25906.
